### DISCUSSION TITLE
Relative URIs don’t work in annotated playlists

### DISCUSSION THREAD

**Discussion by Moonbase59** [savonet/liquidsoap#3743] [User Support]
**Describe the bug**
I have all my playlists using relative URIs, example `Test.m3u`:

```
../../Other/Jingles/test-5-15-5-15-5s.mp3
```
and they play just fine.

When I try to _annotate_ entries, example: `Test-processed.m3u8`:
```
#EXTM3U
annotate:liq_cue_in="5.00",liq_cross_duration="4.69",duration="44.99",liq_amplify="0.50dB":../../Other/Jingles/test-5-15-5-15-5s.mp3
```

I get an `Nonexistent file or ill-formed URI "../../Other/Jingles/test-5-15-5-15-5s.mp3"!` error.

**To Reproduce**
```
uri = "~/Musik/Playlists/Radio/Test-processed.m3u8"
radio = playlist(uri)

def live_aware_crossfade(old, new) =
    log(label="crossfade", "Using custom crossfade")
    cross.smart(old, new, fade_in=0.1, fade_out=2.5, margin=8.)
end

radio = cross(duration=3.00, width=2.00, live_aware_crossfade, radio)

radio = mksafe(radio)

output.ao(radio)
```

**Expected behavior**
Relative URIs working with annotated playlist files.

**Version details**

- OS: Linux Mint 21.3
- Version: Liquidsoap 2.2.4-1

**Install method**
.deb

**Comment by vitoyucepi** [comment]
Hi @Moonbase59,
Could you write absolute paths for these files?

> `~/Musik/Playlists/Radio/Test-processed.m3u8`
> `../../Other/Jingles/test-5-15-5-15-5s.mp3`

Also, what's the current working directory for the liquidsoap script?

**Comment by Moonbase59** [comment]
Can’t convert everything, because the same playlists are used elsewhere…

But if I convert one (for testing) manually, they work _sometimes_. My 2-entry test playlist works:

```
#EXTM3U
annotate:liq_cue_in="5.00",liq_cross_duration="4.69",duration="44.99",liq_amplify="0.50dB":/home/matthias/Musik/Other/Jingles/test-5-15-5-15-5s.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="12.260",duration="432.260",liq_amplify="-7.000dB":/home/matthias/Musik/Tagged/Verest/Verest - [compilations]/Verest - Unity.mp3
```

but a 63-entry file doesn’t:
```
#EXTM3U
annotate:liq_cue_in="0.400",liq_cross_duration="0.070",duration="398.070",liq_amplify="-5.100dB":/home/matthias/Musik/Tagged/Alien Nation/Alien Nation - [compilations]/Alien Nation - Intro _ Gemini 6 - Skysoaring.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="7.720",duration="501.620",liq_amplify="-5.100dB":/home/matthias/Musik/Tagged/Artha/Artha - [compilations]/Artha - Oddity.flac
annotate:liq_cue_in="0.300",liq_cross_duration="3.600",duration="318.300",liq_amplify="-7.700dB":/home/matthias/Musik/Tagged/Avalon/Avalon - Eurasia/Avalon - The Last Call.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="2.350",duration="424.750",liq_amplify="-9.500dB":/home/matthias/Musik/Tagged/Blank & Jones/Blank & Jones - Relax Edition Three/Blank & Jones - Captain of Her Heart (Lounge instrumental).mp3
annotate:liq_cue_in="0.300",liq_cross_duration="0.620",duration="384.020",liq_amplify="-10.000dB":/home/matthias/Musik/Tagged/Blank & Jones/Blank & Jones - Relax Edition Three/Blank & Jones feat. Francis, Mike - Deliver Me (Funk dub).mp3
annotate:liq_cue_in="0.300",liq_cross_duration="1.050",duration="387.050",liq_amplify="-11.200dB":/home/matthias/Musik/Tagged/Blank & Jones/Blank & Jones - Relax Edition Three/Blank & Jones feat. Francis, Mike - Josephine (Balearic House mix).mp3
annotate:liq_cue_in="0.400",liq_cross_duration="11.110",duration="516.410",liq_amplify="-6.400dB":/home/matthias/Musik/Tagged/Brio, Gervay feat. Kalma, Ariel/Brio, Gervay feat. Kalma, Ariel - [compilations]/Brio, Gervay feat. Kalma, Ariel - King of Trance (Didge mix).mp3
annotate:liq_cue_in="0.300",liq_cross_duration="12.710",duration="293.110",liq_amplify="-1.300dB":/home/matthias/Musik/Tagged/Buuren, van, Armin/Buuren, van, Armin - 76 (2003 album, NL)/Buuren, van, Armin feat. Suissa, Justine - Never Wanted This.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="7.090",duration="179.490",liq_amplify="-7.700dB":/home/matthias/Musik/Tagged/Buuren, van, Armin/Buuren, van, Armin - Anthems - Ultimate Singles Collected/Buuren, van, Armin feat. Vayne, Jan - Serenity.flac
annotate:liq_cue_in="0.500",liq_cross_duration="6.930",duration="198.730",liq_amplify="-7.500dB":/home/matthias/Musik/Tagged/Cascada/Cascada - The Offical Remix Album/Cascada - Ready for Love (club mix).mp3
annotate:liq_cue_in="0.300",liq_cross_duration="6.920",duration="396.220",liq_amplify="-5.600dB":/home/matthias/Musik/Tagged/Ceejay/Ceejay - [compilations]/Ceejay - Could This Be Love.flac
annotate:liq_cue_in="0.300",liq_cross_duration="3.090",duration="225.490",liq_amplify="-10.400dB":/home/matthias/Musik/Tagged/Chicane/Chicane - The Place You Can't Remember, The Place You Can't Forget (2018 album, NL)/Chicane & Robinson, Hannah - Gorecki.flac
annotate:liq_cue_in="1.500",liq_cross_duration="1.990",duration="300.790",liq_amplify="2.300dB":/home/matthias/Musik/Tagged/Cold/Cold - [compilations]/Cold - Strobe Light Network.mp3
annotate:liq_cue_in="0.600",liq_cross_duration="2.890",duration="253.590",liq_amplify="-10.900dB":/home/matthias/Musik/Tagged/Cydonia/Cydonia - [compilations]/Cydonia - Midnight Man.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="0.060",duration="280.760",liq_amplify="-2.600dB":/home/matthias/Musik/Tagged/D-Noiser/D-Noiser - [compilations]/D-Noiser - Pump Up The Bass.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="0.590",duration="440.990",liq_amplify="-8.600dB":/home/matthias/Musik/Tagged/Dark Nebula/Dark Nebula - The 8th Sphere (2003 album, AU)/Dark Nebula - Machine of Lum the Mad.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="3.630",duration="297.530",liq_amplify="-7.700dB":/home/matthias/Musik/Tagged/Delacroix, Joris/Delacroix, Joris - Night Visions (2018 album, XW)/Delacroix, Joris & Synapson feat. Schérazade - I Wanna.mp3
annotate:liq_cue_in="0.500",liq_cross_duration="3.730",duration="264.930",liq_amplify="-8.400dB":/home/matthias/Musik/Tagged/Delacroix, Joris/Delacroix, Joris - Night Visions (2018 album, XW)/Delacroix, Joris - Don't Change.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="6.220",duration="264.220",liq_amplify="-7.300dB":/home/matthias/Musik/Tagged/Delacroix, Joris/Delacroix, Joris - Night Visions (2018 album, XW)/Delacroix, Joris - Routine.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="3.000",duration="361.400",liq_amplify="-7.400dB":/home/matthias/Musik/Tagged/Delacroix, Joris/Delacroix, Joris - Night Visions (2018 album, XW)/Delacroix, Joris feat. Ornette - Start the Engine.mp3
annotate:liq_cue_in="0.400",liq_cross_duration="9.230",duration="455.730",liq_amplify="-9.600dB":/home/matthias/Musik/Tagged/Denga & Manus/Denga & Manus - [compilations]/Denga & Manus - Darkness (Vengeance mix).mp3
annotate:liq_cue_in="0.300",liq_cross_duration="7.910",duration="355.410",liq_amplify="-4.700dB":/home/matthias/Musik/Tagged/Dyk, van, Paul/Dyk, van, Paul - In Between (2007 album, TH)/Dyk, van, Paul - Taking the Long Way.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="5.890",duration="353.390",liq_amplify="-4.900dB":/home/matthias/Musik/Tagged/Dyk, van, Paul/Dyk, van, Paul - In Between (2007 album, US)/Dyk, van, Paul - Taking the Long Way.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="3.590",duration="326.290",liq_amplify="-6.800dB":/home/matthias/Musik/Tagged/Dyk, van, Paul/Dyk, van, Paul - Reflections/Dyk, van, Paul - Never Forget.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="0.590",duration="272.790",liq_amplify="-5.900dB":/home/matthias/Musik/Tagged/Dyk, van, Paul/Dyk, van, Paul - Zurdo/Dyk, van, Paul - Te amo.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="0.000",duration="23.900",liq_amplify="-7.400dB":/home/matthias/Musik/Tagged/E Nomine/E Nomine - Finsternis/E Nomine - Die Offenbarung (Interlude).mp3
annotate:liq_cue_in="0.300",liq_cross_duration="2.460",duration="318.660",liq_amplify="-5.600dB":/home/matthias/Musik/Tagged/Eat Static/Eat Static - Back to Earth/Eat Static - Flippity Flippity.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="11.970",duration="287.970",liq_amplify="-5.900dB":/home/matthias/Musik/Tagged/Eat Static/Eat Static - Dead Planet/Eat Static - Daemon Song.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="13.450",duration="378.350",liq_amplify="-9.500dB":/home/matthias/Musik/Tagged/Eat Static/Eat Static - Dead Planet/Eat Static - Irritant.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="8.970",duration="368.070",liq_amplify="-9.600dB":/home/matthias/Musik/Tagged/Eat Static/Eat Static - Dead Planet/Eat Static - Ringlefinch.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="7.490",duration="423.390",liq_amplify="-5.800dB":/home/matthias/Musik/Tagged/Eat Static/Eat Static - Dead Planet/Eat Static - Thoraxian (feat. Mik Singh).mp3
annotate:liq_cue_in="0.300",liq_cross_duration="0.030",duration="162.530",liq_amplify="-8.000dB":/home/matthias/Musik/Tagged/Effective Force/Effective Force - [compilations]/Effective Force - Diamond Bullet.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="1.420",duration="210.620",liq_amplify="-9.200dB":/home/matthias/Musik/Tagged/Gogh, van, Niels/Gogh, van, Niels - Frequenzklang/Gogh, van, Niels - Hit me like a Dirty Dog.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="7.430",duration="198.430",liq_amplify="-8.700dB":/home/matthias/Musik/Tagged/Groove Coverage/Groove Coverage - Greatest Hits/Groove Coverage - Remember.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="0.590",duration="392.590",liq_amplify="-6.400dB":/home/matthias/Musik/Tagged/Haugli & Lizzie Curious/Haugli & Lizzie Curious - [compilations]/Haugli & Lizzie Curious - Your Touch.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="6.180",duration="231.780",liq_amplify="-8.400dB":/home/matthias/Musik/Tagged/Icon of Coil/Icon of Coil - The Soul Is in the Software/Icon of Coil - Disconnect.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="0.050",duration="211.850",liq_amplify="-8.800dB":/home/matthias/Musik/Tagged/Infected Mushroom/Infected Mushroom - Vicious Delicious/Infected Mushroom - Forgive Me.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="4.020",duration="241.420",liq_amplify="-7.600dB":/home/matthias/Musik/Tagged/Jaehn, Felix feat. Thompson, Jasmine/Jaehn, Felix feat. Thompson, Jasmine - [compilations]/Jaehn, Felix feat. Thompson, Jasmine - Ain't Nobody (Loves Me Better).mp3
annotate:liq_cue_in="0.400",liq_cross_duration="8.890",duration="400.090",liq_amplify="-9.900dB":/home/matthias/Musik/Tagged/Jarre, Jean Michel/Jarre, Jean Michel - Remix EP [1] (2015 ep, XE)/Jarre, Jean Michel & 3D - Watching You (JMJ remix extended).flac
annotate:liq_cue_in="0.300",liq_cross_duration="5.830",duration="458.130",liq_amplify="-7.500dB":/home/matthias/Musik/Tagged/Jarre, Jean Michel/Jarre, Jean Michel - Remix EP [1] (2015 ep, XE)/Jarre, Jean Michel & M83 - Glory (16Bit Lolitas remix).flac
annotate:liq_cue_in="0.300",liq_cross_duration="0.000",duration="271.300",liq_amplify="-3.300dB":/home/matthias/Musik/Tagged/Justin K & K-Mixx, DJ/Justin K & K-Mixx, DJ - [compilations]/Justin K & K-Mixx, DJ - Northern Lights.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="4.550",duration="217.750",liq_amplify="-11.400dB":/home/matthias/Musik/Tagged/Moby/Moby - [compilations]/Moby - Why Does My Heart Feel So Bad_ (Ferry Corsten remix).mp3
annotate:liq_cue_in="0.300",liq_cross_duration="2.600",duration="265.900",liq_amplify="-9.900dB":/home/matthias/Musik/Tagged/Negative, Ben/Negative, Ben - [compilations]/Negative, Ben - Milk.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="0.060",duration="253.460",liq_amplify="-4.100dB":/home/matthias/Musik/Tagged/Nostrum/Nostrum - [compilations]/Nostrum - Blow Back.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="6.010",duration="204.010",liq_amplify="-8.000dB":/home/matthias/Musik/Tagged/Prosper/Prosper - Broken Door/Prosper - Broken Door.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="2.420",duration="525.020",liq_amplify="-4.000dB":/home/matthias/Musik/Tagged/Roth, Martin/Roth, Martin - An Analog Guy in a Digital World (2016 single, XW)/Roth, Martin - An Analog Guy in a Digital World.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="3.630",duration="438.830",liq_amplify="-7.700dB":/home/matthias/Musik/Tagged/Sakin, DJ/Sakin, DJ - [compilations]/Sakin, DJ - Everything Is Possible (Memory edit).mp3
annotate:liq_cue_in="0.300",liq_cross_duration="2.720",duration="234.520",liq_amplify="-9.800dB":/home/matthias/Musik/Tagged/Schiller/Schiller - Day and Night/Schiller - Miles and Miles.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="11.520",duration="280.420",liq_amplify="-7.700dB":/home/matthias/Musik/Tagged/Schiller/Schiller - Desire/Schiller with Roth, Jette von - Black.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="7.910",duration="290.610",liq_amplify="-7.700dB":/home/matthias/Musik/Tagged/Schiller/Schiller - Sehnsucht/Schiller mit Roth, Jette von - Black.flac
annotate:liq_cue_in="0.300",liq_cross_duration="3.000",duration="240.300",liq_amplify="-8.100dB":/home/matthias/Musik/Tagged/Schiller/Schiller - Tag und Nacht/Schiller mit Roth, Jette von - Der Tag ... Du bist erwacht.flac
annotate:liq_cue_in="0.300",liq_cross_duration="2.680",duration="388.980",liq_amplify="-9.300dB":/home/matthias/Musik/Tagged/Scorb/Scorb - Scorb (2003 album, GB)/Scorb - Oracle.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="24.070",duration="113.770",liq_amplify="-0.600dB":/home/matthias/Musik/Tagged/Smoke Sign/Smoke Sign - Reincarnate (2014, XW)/Smoke Sign - Boundless.flac
annotate:liq_cue_in="0.300",liq_cross_duration="1.500",duration="193.200",liq_amplify="-7.700dB":/home/matthias/Musik/Tagged/Sphera/Sphera - [compilations]/Sphera - Human.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="8.390",duration="473.690",liq_amplify="-8.100dB":/home/matthias/Musik/Tagged/Star-X/Star-X - Prepare to Fly/Star-X - Shooting Stars.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="3.340",duration="324.040",liq_amplify="-5.500dB":/home/matthias/Musik/Tagged/Sun Spot/Sun Spot - Psychedelic Visions (2021 album, XW)/Sun Spot - How Weird It Is... (original mix).mp3
annotate:liq_cue_in="0.300",liq_cross_duration="4.630",duration="270.230",liq_amplify="-10.300dB":/home/matthias/Musik/Tagged/T & Sugah/T & Sugah - [compilations]/T & Sugah - Give Me.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="9.750",duration="380.050",liq_amplify="-1.600dB":/home/matthias/Musik/Tagged/Taklamakan/Taklamakan - [compilations]/Taklamakan - Taklamakan.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="0.000",duration="365.990",liq_amplify="-8.000dB":/home/matthias/Musik/Tagged/Tomcraft/Tomcraft - For the Queen/Tomcraft - Thomas of Bavaria (feat. President Bongo From Gus Gus).mp3
annotate:liq_cue_in="0.300",liq_cross_duration="0.050",duration="127.450",liq_amplify="-8.300dB":/home/matthias/Musik/Tagged/Top, Emmanuel/Top, Emmanuel - Best Of (2002 album, compilation, FR)/Top, Emmanuel - DJ Set E.T. (9).mp3
annotate:liq_cue_in="0.300",liq_cross_duration="0.010",duration="203.310",liq_amplify="-7.300dB":/home/matthias/Musik/Tagged/True Love/True Love - [compilations]/True Love - Breath of Stars.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="12.260",duration="432.260",liq_amplify="-7.000dB":/home/matthias/Musik/Tagged/Verest/Verest - [compilations]/Verest - Unity.mp3
annotate:liq_cue_in="0.300",liq_cross_duration="11.170",duration="432.770",liq_amplify="-8.400dB":/home/matthias/Musik/Tagged/e-mantra/e-mantra - Arcana (2009 album, BE)/e-mantra - Beyond the Boreas.flac
```

The latter one only generates silence.

The current dir is `~/media/videostream` (where my LS script is), the playlists are in `~/Musik/Playlists/Radio/`, and referred originally to `../../Tagged/something`, which I have now manually changed to `/home/matthias/Musik/Tagged/something`.

Relative paths should normally be resolved with respect to the location of the _playlist file_.

**Comment by vitoyucepi** [comment]
I asked you to manually resolve files for reproduction purposes only.

I can confirm that liquidsoap resolves files in annotated playlists relative to CWD, not the playlist file itself.

**Comment by Moonbase59** [comment]
No offense. Just tried to explain that these are on a NAS and mounted from many different machines, on different  mount points, so relative paths were the choice. Of course I did convert a few just to test. :-)

Think it would be possible to fix this, i.e. have them resolve to the location of the playlist, much like "normal" playlists work? Thinking of embedded Liquidsoap, like in AzuraCast, it might not always be possible to change the CWD to the location of the playlist…

**Comment by toots** [comment]
From a system point of view, I don't think that it is reasonable to expect much from relative URIs. 

Your assumption that it should be referering to the playlist file is not the usual one. in fact. In most situations, relatives URIs are resolved relative the _current working directory_ (CWD) which is the directory from where the user is typing its command.

In general it cannot be expected that all URIs come from a single playlist. URIs can for instance be pushed individually to a `request.queue`. Therefore, given a URI of the form: `annotate:<metadata>:../../bla.mp3` there is no generic notion of playlist attached to it.

In some specific simple case, such as a simple text playlist with relative URIs, it might have been possible to pre-process each URI to resolve the relative URI based on the base playlist. However, as soon at the URIs start being more complex such as one with `annotate`, then getting an actual file path requires applying the various required protocol resolutions to that URIs.

When that happens, the URI is being processed through a generic protocol resolution making it impossible to resolve relative URIs in a consistent way except from using the CWD.

Lastly, the notion of current working directory is essentially a terminal notion, relevant from the perspective of a user being logged into a terminal.

When setting up an automated system, it is not wise to rely on this notion.

Typically, a daemonization utility such as `systemd` will most likely change the path from where the script is executed. In fact, the first step of daemonizing removes all traces of shell/terminal, removing TTYs, etc etc. I would expect the current working directory to be gone by then or to be a trivial location because of that and for security reasons.

In conclusion, if you expect your system to be robust, I would absolutely recommend _against_ using relative URIs anywhere.

It seems to me that your initial remark is because of a lack of referential consistency: a URI with relative path is resolved relative to a playlist when the URI is just the path but not when it is prefixed. As explained above, this is a discrepancy in the resolving conditions. If anything should be done about on liquidsoap side, I would suggest in this case to _remove_ the simple relative URI pre-processing so as to not confuse the use about what will happen when their simple URIs/playlist start being used in a more comple context with protocols and prefixes.

**Reply by vitoyucepi** [reply]
Hi @toots,
Then the behavior should be consistent in playlists with and without annotations.

> I have all my playlists using relative URIs, example `Test.m3u`:
> ```
> ../../Other/Jingles/test-5-15-5-15-5s.mp3
>  ```
> and they play just fine.
> When I try to annotate entries, example: `Test-processed.m3u8`:
> 
> ```
> #EXTM3U
> annotate:liq_cue_in="5.00",liq_cross_duration="4.69",duration="44.99",liq_amplify="0.50dB":../../Other/Jingles/test-5-15-5-15-5s.mp3
> ```
> I get an `Nonexistent file or ill-formed URI "../../Other/Jingles/test-5-15-5-15-5s.mp3"!` error.

I think the correct relativity is to the playlist file, but I don't know what the industry standard is.

**Reply by Moonbase59** [reply]
In my case, about ~160k songs are on my NAS, which gets NFS-mounted to lots of different machines, users, and softwares. So the only reliable way to ensure _everything_ would find the files was to use relative paths. I know of no audio player or playout system (like Jellyfin, KODI, Logitech Media Server, etc.) that _doesn’t_ base the relative paths to the location of the playlist.

Playlists are generated with several automated programs, plus hand-selected ones, and must be usable for all my infrastructure, so I don’t have much chance. Except maybe yet another preprocessing step which seems unneccessary.

I do understand the argument about mixed data and protocols, though. Most audio players, for instance, have no problems with such "mixed" playlists, like maybe
```
https://some.domain.tld/link/to/stream.mp3
/home/matthias/Musik/Playlists/Radio/../../Tagged/Doors, The/Doors, The - In Concert (1991 album, live, US)/Doors, The - Petition the Lord With Prayer.mp3
/home/matthias/Musik/Tagged/Doors, The/Doors, The - In Concert (1991 album, live, US)/Doors, The - Petition the Lord With Prayer.mp3
../../Tagged/Doors, The/Doors, The - In Concert (1991 album, live, US)/Doors, The - Petition the Lord With Prayer.mp3
```

Imagining the annotation being "stripped off", I cannot really see why the two variants behave differently.

**Reply by toots** [reply]
As I said -- once inside the protocol processing pipeline, there is no trace of the original playlist.

Generally speaking, establishing proper boundaries and isolating different elements from each other is the best way to make a system robust.

In your case, I would suggest implementing your own `moonbase:` protocol and using it to abstract away the issue of where the files are located in each installation. Something like this:

```liquidsoap
# This could be part of a config.liq file.
nfs_mount_point = "..."

# Resolve relative urls using the given local mount prefix
def moonbase_proto(~rlog, ~maxtime=_, arg) =
  if string.contains(prefix=".", arg) then
    url = path.concat(nfs_mount_point, arg)
    rlog("Resolving relative url #{arg} to #{url}")
    url
  else
    arg
  end
end

protocol.add(
  "moonbase",
  moonbase_proto,
  syntax="moonbase:path",
  doc="Resolve nfs-based relative paths".
)
```

Then you can do for instance:
```
playlist(
  prefix="annotate:foo=\"bar\":moonbase:",
  ...
)
```

This way, your whole system is more modular and explicit.

**Reply by MaPePeR** [reply]
Couldn't you strip all the protocols from the playlist entry and make the path absolute when you parse the playlist and then re-add the protocols?
Liquidsoap knows all protocols, so it should be able to remove them without executing them.

The `playlist_parser.get_file` function has both the playlist location and the relative path and already does the relative to absolute conversion.
It just needs to be protocol aware:
https://github.com/savonet/liquidsoap/blob/0d8e6eeea2b2b1d3c1e0ccdceea82fbc12c509fa/src/core/playlist_parser.ml#L76-L83

**Reply by toots** [reply]
Well, this could be possible but this also assumes that we can detect with certainty that the last element in the chain is expected to be file and is a relative path.

In the most general sense, anything after `protocol_name:` should be considered private data from a protocol perspective. Again, it's much more clear that, when applying playlist `prefix`, everything before that is actually a file, so tailing the prefix with a ad-hoc protocol would work.

I might design something for it that would more systematic. hold on.

**Reply by MaPePeR** [reply]
I think it would be great if there would be a separate Syntax for protocol arguments and file/url it is acting on, so liquidsoap can always split it.

Something like `annotate(liq_cross_duration="9000"):filepath`.

I already ran into that problem, when I wanted to use my prefix-protocol with a `m3u` playlist that contained metadata and I had to regex away the generated `annotate:` to get to the file path. Not great.

**Reply by MaPePeR** [reply]
Thinking about it more: This also means, that the order of execution of protocols is kind of weird right now?
It feels to me like protocol stacking is kind of encouraged, but they evaluate "left to right" instead of "from inside to outside"/"right to left" as might be expected.

E.g. you can't do something like `a_proto:b_proto:` where `b_proto` resolves to paths in some magic way (like basically _every_ other protocol: `s3`, `http`, `say`, ...) and `a_proto` generates metadata for resulting paths (e.g. `replaygain` or custom protocols). It might work if `a_proto` could resolve `b_proto` itself to get to the filename to then generate metadata? But that's something the language should take care of - at least in my opinion. And I don't think there is a function to do that available?

Something like `playlist(prefix="replaygain:")` probably also doesn't work for any playlist that contains metadata, because it cannot handle the automatically added `annotate:`?

**Reply by Moonbase59** [reply]
"Inner to outer" sounds right. I just have this problem using my `autocue:` protocol in lines like these:
```
autocue:annotate:title="In the Minds of the Young",artist="Pretty Maids",duration="253.00",song_id="b57498b32916bc223f4b0ea11e9b01eb",media_id="282924",playlist_id="234":media:Tagged/Pretty Maids/Pretty Maids - Stripped/Pretty Maids - In the Minds of the Young.mp3
```
"Stackable" would really be nice, but as it is, to run an external binary on the files, I have to resolve AzuraCasts `media:` protocol myself. Oh well, and also disassemble and reassemble the `annotate:` values.
